### PR TITLE
Add garbage protos

### PR DESCRIPTION
**Pull request by blowmage**
These protos are a way to cover all edge cases not covered in the other protos. They can be used to test all field types, all RPC method types, all annotation, etc.

### CODE REVIEW

**Comment by blowmage** [comment]
This PR doesn't include 1) the changes necessary to get the updated annotations working properly 2) updates to the cloud generator to get it to build the garbage protos 3) updated helper tests using the additional fields these protos provide. But the protos do include the updated annotations and are compiling without warning. I expect the garbage protos to be added to over time as new areas for test coverage are identified.

File: shared/protos/garbage/garbage.proto [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+import "google/api/annotations.proto";
+import "google/api/client.proto";
+import "google/api/resource.proto";
+import "google/longrunning/operations.proto";
+import "google/protobuf/timestamp.proto";
+
+option (google.api.client_package) = {
+  title: "Testing Garbage"
+  namespace: ["Google", "Cloud"]
+  version: "v1"
```
Comment (jbolinger) [inline shared/protos/garbage/garbage.proto]: Same. I'd prefer to use something strange that does not look like a version or follow a typical version regex.

File: shared/protos/garbage/garbage.proto [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+};
+
+// Endless trash
+service GarbageService {
+  // The address where this API is typically served.
+  option (google.api.default_host) = "garbage.googleapis.com";
```
Comment (jbolinger) [inline shared/protos/garbage/garbage.proto]: Same. More random trash and less Google references.

File: shared/protos/garbage/garbage.proto [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+  // The OAuth scopes needed to make requests to this API.
+  option (google.api.oauth_scopes) = "https://cloud.google.com/auth/cloud-platform,"
```
Comment (jbolinger) [inline shared/protos/garbage/garbage.proto]: Same. Need more trash here.

File: shared/protos/garbage/garbage.proto [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    };
+    // Once the operation is complete, the ultimate result
+    // will be a LongRunningGarbageResponse.
+    option (google.longrunning.operation_info) = {
+      response_type: "google.garbage.v1.LongRunningGarbageResponse"
+      metadata_type: "google.garbage.v1.LongRunningGarbageMetadata"
+    };
+  }
+
+  // This method split the given content into words and will pass each word back
```
Comment (jbolinger) [inline shared/protos/garbage/garbage.proto]: Why do these methods have the same comments as showcase (copy/paste)? Some comments are fine but we probably shouldn't use the same one here.

File: shared/protos/garbage/garbage.proto [hunk 5]
```diff
@@ -0,0 +1,234 @@
+syntax = "proto3";
+
+package google.garbage.v1;
```
Comment (jbolinger) [inline shared/protos/garbage/garbage.proto]: Let's use a namespace that doesn't include `google` and doesn't end in a version.

File: shared/protos/garbage/garbage.proto [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+message SimpleGarbageItem {
+  string garbage = 1 [
+    (google.api.resource_reference) = "SimpleGarbage"];
+  oneof item {
+    string a = 2;
+    int32  b = 3;
+  }
+}
+
+message TypicalGarbage {
```
Comment (jbolinger) [inline shared/protos/garbage/garbage.proto]: Maybe add a nested type or two for completeness?

**Comment by quartzmo** [comment]
Now that I see this, I'm wondering what makes it different from gapic-showcase `EchoClient`?

**Comment by jbolinger** [comment]
@quartzmo it's probably not so different as we get started but it will be once we really get going (and testing). Showcase is sort of a normal API and we'll want a proto that we can put all sorts of strange things in to test edge cases, etc. that we can't put in showcase. Hence, this garbage :)

File: shared/output/gapic/templates/garbage/google/garbage/v1/garbage_service.rb [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+require "pathname"
+
+require "google/gax"
+require "google/gax/operation"
+require "google/longrunning/operations_client"
+
+require "google/cloud/garbage_service/v1/cloud_garbage_service_pb"
+require "google/cloud/garbage_service/v1/credentials"
+
+module Google
```
[the remark this answers is not in the document]
Comment (blowmage) [reply shared/output/gapic/templates/garbage/google/garbage/v1/garbage_service.rb]: One is the file to compare for tests, one is the output from generating the binary file. I have a comment in the binary file rake task to send the output to `/dev/null` instead.

File: gapic-generator-cloud/Rakefile [hunk 8]
```diff
@@ -43,6 +43,7 @@ namespace :gen do
   task :templates do
     generate_cloud_templates "speech"
     # generate_cloud_templates "vision"
+    # generate_default_templates "garbage"
```
[the remark this answers is not in the document]
Comment (blowmage) [reply gapic-generator-cloud/Rakefile]: The cloud generator currently breaks when trying to generate anything other than speech, because so much of it is faked out and specific to speech. The gapic generator is able to build all the inputs, so it isn't commented out.

File: shared/protos/garbage/garbage.proto (hunk 6 above)
Comment (blowmage) [reply shared/protos/garbage/garbage.proto]: How do you mean? Isn't `GarbageMap` a nested type?

**Comment by blowmage** [comment]
@quartzmo There are many things that this has that showcase does not, such as using all available field types, defining `oneof` fields, using `resource`, `resource_reference`, and `resource_set` annotations, etc...

File: shared/protos/garbage/garbage.proto (hunk 6 above)
Comment (jbolinger) [reply shared/protos/garbage/garbage.proto]: No, I mean something like:

```proto
message Foo {
    message Bar {
        // ....
    }
    // ....
}
```

It's not something we need to do right now, but at some point we will want to make sure the generator can handle nested types correctly.

File: shared/protos/garbage/garbage.proto [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+import "google/api/annotations.proto";
+import "google/api/client.proto";
+import "google/api/resource.proto";
+import "google/longrunning/operations.proto";
+import "google/protobuf/timestamp.proto";
+
+option (google.api.client_package) = {
+  title: "Testing Garbage"
+  namespace: ["Google", "Cloud"]
```
[the remark this answers is not in the document]
Comment (blowmage) [reply shared/protos/garbage/garbage.proto]: Trashed it up.

File: shared/protos/garbage/garbage.proto (hunk 1 above)
Comment (blowmage) [reply shared/protos/garbage/garbage.proto]: Trashed it up.

File: shared/protos/garbage/garbage.proto (hunk 2 above)
Comment (blowmage) [reply shared/protos/garbage/garbage.proto]: Trashed it up.

File: shared/protos/garbage/garbage.proto (hunk 3 above)
Comment (blowmage) [reply shared/protos/garbage/garbage.proto]: Trashed it up.

File: shared/protos/garbage/garbage.proto (hunk 4 above)
Comment (blowmage) [reply shared/protos/garbage/garbage.proto]: Removed.

File: shared/protos/garbage/garbage.proto (hunk 5 above)
Comment (blowmage) [reply shared/protos/garbage/garbage.proto]: Trashed it up.

File: shared/output/gapic/templates/garbage/google/garbage/v1/garbage_service.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (blowmage) [reply shared/output/gapic/templates/garbage/google/garbage/v1/garbage_service.rb]: These files are now removed.

File: shared/protos/garbage/garbage.proto (hunk 6 above)
Comment (blowmage) [reply shared/protos/garbage/garbage.proto]: Oooohhh, I didn't even realize we could do that...

Comment (blowmage) [reply shared/protos/garbage/garbage.proto]: Added.

**Comment by blowmage** [comment]
Endless trash!

**Comment by blowmage** [comment]
I didn't add the `google.api.method_signature` annotations because we already discussed changing what is accepted to the client methods. It would be nice to allow users to provide either the request object as a positional argument, or the request object values as named arguments. Because of this, we won't be generating method signatures using these annotations. But, we could generate code examples using them.

What do you all think? Is it worth adding `google.api.method_signature` annotations to the garbage proto?

**Comment by jbolinger** [comment]
I would vote for adding a `method_signature` somewhere even if we completely ignore it in the generator. 

It'll serve as explicit documentation that this annotation has no effect and since it'll be commonly found in the input it wouldn't be bad to have some simple test that shows it doesn't do anything. Not a high priority item by any means though.

**Comment by blowmage** [comment]
K, added some `method_signature` and `field_behavior` annotations.

**Comment by jbolinger** [comment]
@blowmage the extra content looks good. Are you planning to do anything more here?

**Comment by blowmage** [comment]
I don't have anything else for this PR.
